### ISSUE TITLE
Download .m3u8 links directly

### ISSUE THREAD

**Issue by Liz-chan** [Franiac/TwitchLeecher#624]
I think a useful feature would be if you could add a link to an .m3u8 file directly if you already have it copied instead of going through the standard process, since that would save time in some instances and also allow you to save sub-only VODs if you've been able to get the direct stream link. At the moment the "URLs" tab in TL only works for Twitch video links, and I think there would be the perfect place to add it.

**Comment by Franiac** [comment]
How would a normal user get a link to a Twitch VOD m3u8 file?

**Comment by Liz-chan** [comment]
Well, aside from using the network inspector in the browser, you can also use a tool like [this](https://github.com/TwitchRecover/TwitchRecover) in order to get direct .m3u8 links for sub-only or unlisted VODs.

**Comment by Zurlon** [comment]
For what's it worth I absolutely second this, if there's one feature I could choose to be added it's this. I know the aim of the project is to be simple so even if this feature was somewhat hidden and not openly displayed it would still be hugely useful.

**Comment by Zurlon** [comment]
Since you already have a text box to enter Twitch URLs into, would it be feasible to allow (advertised or not) a single M3U8 link to be entered into that box, and download it as-is? (with just file location/name as the options)

**Comment by Franiac** [comment]
My main concern is that downloading a m3u8 playlist has nothing to do with Twitch. It's as easy as that. And I think you guys are not taking this into account. A playlist can be hosted anywhere on the internet. I don't see why this functionality should be part of **TWITCH** Leecher. A Playlist does not provide a title, resolution, date, thumbnail... nothing but a total length. So the download UI would also be totally emtpy because yea... It has absolutely nothing to do with a Twitch video, it's just a plain and simple file download. But TL is not a file downloader, it's a Twitch VOD downloader.

If you dumb it down to what you guys want you end up with 2 simple tasks:

1. Download all the files in the playlist
2. Merge the files into a single one

This could even be done with a batch script in half an hour and there a hunderets of existing playlist downloaders, so tools already exist. Just google "m3u8 playlist downloader" and choose one.

**Comment by Liz-chan** [comment]
Maybe it would be a better idea to add support for sub-only VODs with the code from [Twitch Recover](https://github.com/TwitchRecover/TwitchRecover) or [TwitchSubVod](https://github.com/Alissonsleal/TwitchSubVod) then?

The main benefit that I was looking at for using TL to download .m3u8 links is that it's already aware of the muted segments of VODs (while other .m3u8 downloaders aren't because they're not made for twitch), and because renaming and queuing would be incredibly useful and something I can't see a current tool doing. If you know of something similar that has easy renaming and queuing I'd love to know.

**Comment by Franiac** [comment]
@Liz-chan I think you got something wrong there. TL is not aware of anything. TL downloads the playlist and downloads each file within it as is. What would be different if any other tool downloaded that exact same playlist?
If Tiwtch mutes parts, it actually changes the playlist. There's no difference between what tool downloads it or plays it back.

TL had sub-only support for a very long time before Twitch fucked it up. I removed it because at the time it was not possible to get it working again. Maybe things have changed, but nowadays I face the problem that I don't know where to steal that time from to work on TL. TL came out in 2016 and since then a lot has changed (for me personally).

You see all the open issues and it hurts a little bit that nothing is going forward anymore. So yea... honestly I don't know what to do with TL at the moment. I am keeping it alive basically. That is also the explanation why I don't implement stuff that simply does not need to be implemented at the moment.

**Comment by Zurlon** [comment]
FWIW the only reason I would like to have the feature added is because it seems like implementing the functionality that TwitchRecover etc has is not within the scope of, as you say, just maintaining the project.

So my thought was, allow us to use the programs with working sub-only VOD code to get the m3u8 links, and let us download it using your program so those downloads can be integrated with other TL downloads in the queue, so as not to have 2 high bandwidth downloads occuring at once.

I accept your point about opening it up to any m3u8 links, surely you could check the URL to see if it's a twitch URL to keep the purity (which I respect) of TL as a Twitch only downloader?

**Comment by Liz-chan** [comment]
Oh it seems that was my mistake, since I just downloaded an .m3u8 directly right now and it had the -muted.ts parts appended to it already, so I'm not sure why Twitch Recover has something where it "unmutes" the VOD. Apparently it was for making it work in VLC and I guess that's the case since I just tried using the direct link from Twitch right now and the video freezes during the muted segments, but I guess it might not matter when you're downloading it.

Ah I see, it's alright if you don't have enough time to work on TL that much anymore, it's still been a huge help to me anyway even if I have to do some things manually concerning sub-only VODs.

**Comment by Franiac** [comment]
I mean it may be an unpopular opinion but if the guy of TwitchRecover throws out 2.0 and has everythin implemented from his feature table I can basically stop working on TL because his tool has everything TL has and all the fancy stuff on top. I mean I am even in his "Credits" so I guess I helped him with my source code, too.

Software comes and goes... maybe it's time for the next best Twitch downloader?

**Comment by Liz-chan** [comment]
Well right now 2.0 is only in alpha and the features are pretty barebones (I only use it for downloading sub-only VODs and finding deleted clips), it doesn't have have a GUI or even queuing or mass downloading so you have to do everything manually and it's a very slow process. For everything else I use TL because it's mature, has a GUI, and everything else pretty much except being able to download sub-only VODs.

**Comment by Franiac** [comment]
@Liz-chan Ok, that was just a shot in the blue. I never tried TwitchRecover but it seemed to spit big words on it's GitHub page so I thought there would something behind it. It's Java... maybe the developer relized how much work a proper Java GUI actually is (even worse than WPF).

Here's a little teaser:

I have implemented a full blown multithreaded download manager (remember issue #46 anyone???). It already works... BUUUUUUUT... it's in NodeJS/Electron. I hate WPF and it's stupid GUI framework. And that does not help with motivating me to implement anything in TL right now.

So I thought I could do a proper re-write of TL with all the open enhancements as an Electron/Angular dektop app. That would be completely insane and very extensible. Proper CLI, even better GUI.

Theres one problem to it... I don't want to make the download manager open source. And people may hate that idea.

**Comment by Liz-chan** [comment]
I think NodeJS/Electron would work just fine, and it's a lot easier for making GUIs since it's just HTML and CSS. I think a lot of people don't like Electron because of the RAM usage and because a lot of developers use it for simple apps, but as long as Electron TL is optimized I wouldn't have any issues with it. I think Discord is a great Electron app, it's pretty fast and the resource usage is fine for the most part, and although a native version would be a lot less resource intensive, I understand that it would be a lot harder to develop for (also harder to mod as well.

I don't really mind if the download manager was closed source, a lot of us use closed source software every day and I don't believe you have any ill intentions so.

**Comment by Franiac** [comment]
i am not going to implement this in Twitch Leecher .NET. I put it on my features list for the Electron version.
